File: src/main/java/com/indiepost/controller/api/admin/AdminStaticPageController.java
package com.indiepost.controller.api.admin;

import com.indiepost.dto.DeleteResponse;
import com.indiepost.dto.StaticPageDto;
import com.indiepost.enums.Types;
import com.indiepost.service.StaticPageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.*;

/**
 * Created by jake on 17. 3. 5.
 */
@RestController
@RequestMapping(value = "/api/admin/pages", produces = {"application/json; charset=UTF-8"})
public class AdminStaticPageController {

    private final StaticPageService staticPageService;

    @Autowired
    public AdminStaticPageController(StaticPageService staticPageService) {
        this.staticPageService = staticPageService;
    }

    @RequestMapping(method = RequestMethod.GET)
    public Page<StaticPageDto> getList(@RequestParam String status, Pageable pageable) {
        return staticPageService.find(Types.PostStatus.valueOf(status.toUpperCase()), pageable);
    }

    @RequestMapping(method = RequestMethod.POST)
    public Long savePage(@RequestBody StaticPageDto staticPageDto) {
        return staticPageService.save(staticPageDto);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public StaticPageDto getPage(@PathVariable Long id) {
        return staticPageService.findById(id);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.PUT)
    public void updatePage(@PathVariable Long id, @RequestBody StaticPageDto staticPageDto) {
        if (!staticPageDto.getId().equals(id)) {
            return;
        }
        staticPageService.update(staticPageDto);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
    public DeleteResponse delete(@PathVariable Long id) {
        staticPageService.deleteById(id);
        return new DeleteResponse(id);
    }

}

File: src/main/kotlin/com/indiepost/controller/admin/AdminStaticPageController.kt
package com.indiepost.controller.admin

import com.indiepost.dto.DeleteResponse
import com.indiepost.dto.StaticPageDto
import com.indiepost.dto.post.BulkStatusUpdateDto
import com.indiepost.enums.Types
import com.indiepost.service.StaticPageService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.web.bind.annotation.*

import javax.inject.Inject
import javax.validation.ValidationException

/**
 * Created by jake on 17. 3. 5.
 */
@RestController
@RequestMapping(value = ["/admin/pages"], produces = ["application/json; charset=UTF-8"])
class AdminStaticPageController @Inject
constructor(private val staticPageService: StaticPageService) {

    @GetMapping
    fun getList(@RequestParam status: String, pageable: Pageable): Page<StaticPageDto> {
        return staticPageService.find(Types.PostStatus.valueOf(status.toUpperCase()), pageable)
    }

    @PostMapping
    fun savePage(@RequestBody staticPageDto: StaticPageDto): Long? {
        return staticPageService.save(staticPageDto)
    }

    @GetMapping("/{id}")
    fun getPage(@PathVariable id: Long): StaticPageDto? {
        return staticPageService.findById(id)
    }

    @PutMapping("/{id}")
    fun updatePage(@PathVariable id: Long, @RequestBody staticPageDto: StaticPageDto) {
        if (staticPageDto.id != id) {
            throw ValidationException("REST path variable ID and AdminPostRequestDto::id does not match.")
        }
        staticPageService.update(staticPageDto)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): DeleteResponse {
        staticPageService.deleteById(id)
        return DeleteResponse(id)
    }

    @PutMapping("/_bulk")
    fun bulkUpdate(@RequestBody dto: BulkStatusUpdateDto) {
        val ids = dto.ids
        if (ids == null || ids.isEmpty()) {
            throw ValidationException("BulkStatusUpdateDto::ids cannot be null or empty.")
        }
        val status = Types.PostStatus.valueOf(dto.status.toUpperCase())
        staticPageService.bulkUpdateStatus(ids, status)
    }

    @DeleteMapping("/_bulk")
    fun bulkDelete(@RequestBody dto: BulkStatusUpdateDto) {
        val ids = dto.ids
        if (ids == null || ids.isEmpty()) {
            throw ValidationException("BulkStatusUpdateDto::ids cannot be null or empty.")
        }
        staticPageService.bulkDeleteByIds(ids)
    }

    @DeleteMapping("/_trash")
    fun emptyTrash() {
        staticPageService.emptyTrash()
    }
}
